libcloud/backup/drivers/gce.py

```diff
@@ -0,0 +1,478 @@
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    'GCEBackupDriver'
+]
+
+from libcloud.utils.iso8601 import parse_date
+from libcloud.backup.base import BackupDriver, BackupTargetRecoveryPoint,\
+    BackupTargetJob, BackupTarget
+from libcloud.backup.types import BackupTargetType, BackupTargetJobStatusType
+
+from libcloud.common.google import GoogleResponse, GoogleBaseConnection
+
+API_VERSION = 'v1'
+DEFAULT_TASK_COMPLETION_TIMEOUT = 180
+
+
+class GCEResponse(GoogleResponse):
+    pass
+
+
+class GCEConnection(GoogleBaseConnection):
+    """
+    Connection class for the GCE driver.
+
+    GCEConnection extends :class:`google.GoogleBaseConnection` for 2 reasons:
+      1. modify request_path for GCE URI.
+      2. Implement gce_params functionality described below.
+
+    If the parameter gce_params is set to a dict prior to calling request(),
+    the URL parameters will be updated to include those key/values FOR A
+    SINGLE REQUEST. If the response contains a nextPageToken,
+    gce_params['pageToken'] will be set to its value. This can be used to
+    implement paging in list:
+
+    >>> params, more_results = {'maxResults': 2}, True
+    >>> while more_results:
+    ...     driver.connection.gce_params=params
+    ...     driver.ex_list_urlmaps()
+    ...     more_results = 'pageToken' in params
+    ...
+    [<GCEUrlMap id="..." name="cli-map">, <GCEUrlMap id="..." name="lc-map">]
+    [<GCEUrlMap id="..." name="web-map">]
+    """
+    host = 'www.googleapis.com'
+    responseCls = GCEResponse
+
+    def __init__(self, user_id, key, secure, auth_type=None,
+                 credential_file=None, project=None, **kwargs):
+        super(GCEConnection, self).__init__(user_id, key, secure=secure,
+                                            auth_type=auth_type,
+                                            credential_file=credential_file,
+                                            **kwargs)
+        self.request_path = '/compute/%s/projects/%s' % (API_VERSION,
+                                                         project)
+        self.gce_params = None
+
+    def pre_connect_hook(self, params, headers):
+        """
+        Update URL parameters with values from self.gce_params.
+
+        @inherits: :class:`GoogleBaseConnection.pre_connect_hook`
+        """
+        params, headers = super(GCEConnection, self).pre_connect_hook(params,
+                                                                      headers)
+        if self.gce_params:
```

Comment:
I'm OK with this for now, but when possible it's better to avoid mutating class level variables.

Mutating class level variables could cause issues when using something like eventlet on a single driver instance, but then again, our documentation already says that if you want to use something like eventlet / gevent you should not share a driver instance and use one driver instance per green thread.